### DISCUSSION TITLE
Better naming of row related commands

### DISCUSSION THREAD

**Discussion by nome** [nushell/nushell#13779] [General]
Apologies if it's a little too late into the development of nushell to suggest such broad breaking changes, but I more or less stumbled across this issue today, and thought I'd share my thoughts just in case.

## the problem

I had some difficulty today finding a way to drop a specific row from a table (turns out, it's `drop nth`), so I tried to get an overview of the various row and column related filter commands. Here's what I came up with:

| part of table | keep | discard | add |
|-|-|-|-|
| specific rows | get | drop nth | - |
| range of rows | range | drop nth | - |
| first n rows | first/take, take while, take until | skip, skip while, skip until | prepend |
| last n rows | last | drop | append |
| every nth row | every | every --skip | - |
| specific columns | select | reject | insert plus move |
| last n columns | - | drop column | insert |

Some oddities in function naming immediately stand out: `drop nth` is (apart from `drop column` and the while/until family) the only subcommand; and it's also - in comparison to other nushell builtins - rather cryptic. In my understanding, nushell commands tend to favor legibility over brevity. `every --skip` represents the only functionality in this table that doesn't have its own command, but requires a flag. `drop nth` accepts either one or more indices or a range, while the equivalent "keep" functionality is separated into `get` and `range`. `first` and `take` implement almost the same functionality, while [nu philosphy](https://www.nushell.sh/contributor-book/philosophy_0_80.html) states that "There should be one -- and preferably only one -- obvious way to do it." Not to mention that it's a total of 14 keywords just for basic table manipulation (get first take last range every select skip drop reject prepend append insert move).

All in all, I find this set of commands a little hard to memorize. So I tried to come up with some better names.

## a possible solution

| part of table | keep | discard | add |
|-|-|-|-|
| specific rows | get | drop | - |
| range of rows | get | drop | - |
| first n rows | first, take while, take until | drop first, drop while, drop until | prepend |
| last n rows | last | drop last | append |
| every nth row | every | drop every | - |
| specific columns | select | reject | insert plus move |
| last n columns | - | drop column | insert |

This means two fewer keywords to remember (skip and range), as well as more symmetrical command naming betwen the "keep" and "discard" functionalities.

Compared to the current commands, this changes the following:

* integrate `range` functionality into `get` (which currently does not accept ranges)
    * alternatively, add `drop range` command for symmetry with `range`
* rename `drop` to `drop last` for symmetry with `last`
* rename `drop nth` to `drop`
* add `drop every` command which takes over `every --skip` functionality
* rename `skip` to `drop first`
* rename `skip while/until` to `drop while/until`
* merge `take` into `first` (this is a bit of a compromise, since doing it the other way around would eliminate one more keyword; however, in many situations, `first` is the more natural choice, and it's an obvious complement to `last`)

## outlook

You can test drive some of these changes using the following aliases:

```
alias "drop last" = drop
alias drop = drop nth
alias "drop every" = every --skip
alias "drop first" = skip
alias "drop while" = skip while
alias "drop until" = skip until
```

I'm not sure how typical my problems are of beginner issues. A somewhat related discussion on asymmetrical commands is in #10670, though that discussion doesn't mention most of the commands dicsussed above. Let me know what you think.

**Comment by NotTheDr01ds** [comment]
I think I'll probably have more to comment on tomorrow, but in the meantime see #11218 for some similar ideas.

**Reply by nome** [reply]
Thanks, I'll have a look at that one.

**Comment by T3sT3ro** [comment]
I find `drop` and `drop nth` very confusing. They could be consolidated into a single function `drop [column] <int|range>` for simplicity.

This is also a good post indicating the confusing amount of similar yet different functions. The simplest way I see it could be improved would be to provide positive/negative filter specifier, for example `pick [!]<int|range>` where `!` would invert the rows/columns based on the range/index or something

Additionally I don't see why the functionality of `reject` couldn't be absorbed by `drop` - the syntax would be the same.
